viewing_party/party.py

```diff
@@ -1,23 +1,239 @@
+        rating (float): Rating given to movie.
+
+    Returns:
+        dict: represents a movie with information about its title, genre, and rating.
+    """
+    list_movies = user_data["watchlist"]
+    for movie in list_movies[:]:
+        if movie["title"] == title:
+            list_movies.remove(movie)
+            user_data["watched"].append(movie)
+            break 
+    return user_data
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    """
+    Return the average rating of all movies in the user's watched list.
+    If the list is empty, return 0.0.
+    """
+    rating_total = 0.0
+    list_movies = user_data["watched"]
+    if not list_movies:
+        return 0.0
+    for movie in list_movies:
+        rating_total += movie["rating"]
+    return rating_total / len(list_movies)    
+
 
+def get_most_watched_genre(user_data):
+    """Finds the most frequently occurring genre in the watched movie list.
 
+    Args:
+        user_data (dictionary):
+            "watched": List of dictionaries that represent movies the user has watched.
+                Movie dictionaries has the following key-value pairs:
+                    title (str)
+                    genre (str)
+                    rating (float)
+
+    Returns:
+        genre (str): Determines the most frequently occurring genre among the dictionaries of watched movies. Returns None if value of "watched" is an empty list.
+    """
+    if not user_data["watched"]:
+        return None
+    
+    genre_frequency = {}
+    for movie in user_data["watched"]:
+        genre_frequency[movie["genre"]] = genre_frequency.get(movie["genre"], 0) + 1
+    
+    most_watched_genre = ""
+    most_watched_genre_count = 0
+    for genre, count in genre_frequency.items():
+        if count > most_watched_genre_count:
+            most_watched_genre = genre
+            most_watched_genre_count = count
+    
+    return most_watched_genre
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    """
+    Return a list of movies that appear in the user's watched list
+    but are not present in any friend's watched list.
+    Each movie is represented as a dictionary with title, genre, and rating.
+    """
+    unique_movies = []
+    user_watched = user_data["watched"]
+    friends = user_data["friends"]
+
+    friends_title = set()
+    for watched_dict in friends:
+        for movie in watched_dict["watched"]:
+            title = movie["title"]
+            if title:
+                friends_title.add(title)
+
+    watched = set()
+    for movie in user_watched:
+        title = movie["title"]
+        if title and title not in friends_title and title not in watched:
+            unique_movies.append(movie)
+            watched.add(title)

```

Comment:
We could make the assumption that a `movie` cannot have a `Falsy` value as a `title` since we explicitly check for that in `create_movie`. Doing so would allow us to remove the `if title` check. 

If we assume the that a user's watched list cannot contain duplicate movies (meaning dictionaries' cannot have `title` keys that are `==` to each other) we could remove the `title not in watched` as well. However, there is no supporting evidence in the codebase for this assumption. You would need to check with other developers (us instructors in this case!).